posts/evix: init (#128)
NotAShelf/notashelf.dev: content/posts/evix.md

---
title: "Nix Evaluation Is a Scheduling Problem"
description: "Why I built evix, and why the interesting part was never about talking to Nix"
date: "2026-06-25"
keywords: ["nix", "rust", "programming"]
---

Most, if not all, Nix tools I build (or at least, consider building) eventually
want to do one little boring thing. They want _a list of derivations an
expression produces_. Not necessarily a build log---I've got [a crate for that]
already and those are relatively trivial to parse already and nor a closure but

> a **a list of jobs**. Those usually require their attribute paths, names,

Reviewer: ```suggestion
already and those are relatively trivial to parse already and not a closure but **a list of jobs**. Those usually require their attribute paths, names,
```

systems, `.drv` paths and outputs to be visible and _if_ you're building
something like Hydra then you need those before you can decide what to build.

> I.e., those are the things a CI system needs before it can decide what changed.

Reviewer: ```suggestion
i.e., those are the things a CI system needs before it can decide what changed.
```

While I haven't quite tried, those are also things I imagine a deployment tool
needs before it can decide whether the graph in front of it is the same

> abomination it saw five minutes.

Reviewer: ```suggestion
abomination it saw five minutes ago.
```

_**sucks**_. And I can tell you that I am not interested in a cache-shaped
_thing_ that's masquerading as one.

For a very long time I've thought that my complaint was that evaluation was slow
(it is) and that the output was annoying to parse (it is). It's roughly the same
kind of experience as having an entire piano fall on top of your head while
taking a stroll but that's not what ruins my day. What annoys me the most is
_not_ having a gravitationally-challenged piano deciding to ruin my evening
stroll but the cost that is the program doing the evaluation had already built
the interesting structure, printed a representation of it, and then died.
Naturally the graph dies with it. Go figure.

When I started calling `jq` (which I'm not a huge fan of) I slowly started to
grow suspicious that the tool I'm using is... wrong. As I've mentioned, I
initially started by slapping `nix-eval-jobs` on top of my problems and I _do
not_ want to lead you to think it is bad software, [^1] but it's pretty much
_not_ what I was looking for. There were still a lot of annoying gaps that would
take too long to fill via traditional means and I _know_ it'd suck to do so

> because C++ sucks. Anyway my point is that it's a relatively good tool for

Reviewer: ```suggestion
because C++ sucks. Anyway, my point is that it's a relatively good tool for
```

> finder for what a Nix job evaluator should find and [evix] still leans on it in

Reviewer: ```suggestion
finding for what a Nix job evaluator should find and [evix] still leans on it in
```

benchmarks because I acknowledge it, and I strive to be _better_ than it.

[^1]: I'm pointing out so that you don't mistakenly think that I'm inherently
    negative to every project I aim to replace. I do have my reservations about
    nix-eval-jobs and some of its development but it would be very easy and very

>     easy to dishonestly tell you evix is good because nix-eval-jobs is bad. The

Reviewer: ```suggestion
    nix-eval-jobs and some of its development but it would be very easy to dishonestly tell you evix is good because nix-eval-jobs is bad. The
```

│  └─ aarch64-linux      attrset
│     ├─ hello           derivation (leaf)
│     └─ git             derivation (leaf)
└─ devShells
   └─ x86_64-linux
      └─ default         derivation (leaf)
```

Behold my artistic skills! The recursion does not live inside a call stack that
one worker owns. It lives in a queue. The coordinator seeds the queue with the
root path. A worker pulls a path, evaluates it, and returns an event. An attrset
event adds more paths to the queue. A derivation event is terminal. An error
event is terminal for that path. The run ends when the queue is empty and no
worker is busy.

Walked as a queue rather than a private recursion, that tree drains roughly like
this, in whatever order the workers happen to answer:

```text
queue: [ .# ]
  .#                            attrset    -> enqueue packages, devShells
  packages                      attrset    -> enqueue packages.x86_64-linux, ...
  packages.x86_64-linux         attrset    -> enqueue ...hello, ...git, ...ripgrep
  packages.x86_64-linux.hello   derivation -> emit; enqueue nothing
  packages.x86_64-linux.git     derivation -> emit; enqueue nothing
  …
empty queue, nothing in flight  → done
```

> [!NOTE]

> > evix does not recurse into every attrset by default. It follows the same basic

Reviewer: ```suggestion
> Evix does not recurse into every attrset by default. It follows the same basic
```

> idea tools like this usually follow: recurse at the root, and recurse into
> attrsets that opt in with `recurseForDerivations`. There is a
> `--force-recurse` option for the times you really do want to kick down every
> door, but the default is _less_ stupid. Nixpkgs has many rooms. Some contain
> jobs. Some contain machinery. Some contain things that make you regret
> curiosity as a concept.

The important invariant, which I call important because it caused many
breakages, is that the answer is a graph keyed by attribute path, not by the
order events happened to arrive. One worker and four workers should produce the
same graph. Naturally, local workers and remote workers should feed the same
accumulator. The event stream is temporal because programs live in time, but the
resulting graph is not supposed to depend on scheduler luck. That is the licence
to use parallelism. [^2]

[^2]: [We used to have good things](https://www.imdb.com/title/tt0097742/).

> It's also worth talking about Evix's benchmark suite. I'm _yet_ to to benchmark

Reviewer: ```suggestion
It's also worth talking about evix's benchmark suite. I'm _yet_ to benchmark
```

    warm query inside the library is a map scan, but the CLI still pays for
    daemon IPC, serialization, and writing output. Full-output queries are not
    magic. Narrow queries are where the warm graph starts to feel like cheating,
    because it is no longer doing evaluation work to answer them.

## Processes, Not Threads

If you have ever watched a Nix evaluation climb past the memory you have and get
killed three-quarters of the way through, taking the whole run with it, this is
the part that concerns you. [^5] Evix workers are subprocesses, not threads. The
host process owns the queue and the warm graph. Each worker owns its Nix
context, store handle, eval state and root value. The boundary is crude in the
way Unix is crude, which is to say it is one of the few crude things in
computing that still works.

[^5]: It concerns me _so much_ in fact, that I have Nix cgrouped to limit how
    much CPU, RAM and IO it can use thanks to Systemd's cgroup accounting.
    Without it, Nix evaluation _regularly_ kills low-end machines that I happen
    to have the misfortune of rebuilding on.

After each attribute, the worker checks its maximum resident set size against a

> configured limit, 4 gbs by default. If the worker is over the line, it returns a

Reviewer: ```suggestion
configured limit, 4 GB by default. If the worker is over the line, it returns a
```

  can also leave a warm session behind.
- `evix query` asks an already warm daemon session for a filtered snapshot.
- `evix diff` re-evaluates and compares against the warm graph.
- `evix watch` evaluates a local file or flake and emits diffs as watched inputs
  change.
- `evix worker` starts a remote worker service.
- `evix daemon` starts the Unix-socket daemon explicitly, if you do not want the
  CLI to pretend all of this happens by divine intervention.

[Manic Systems]: https://github.com/manic-systems

So my point is that Evix is _not merely a faster way to print the same list_. It
is a way to keep the evaluated graph resident long enough that follow-up
questions become ordinary program operations. It also isn't a Nix eval wrapper.
Evix was specifically designed for [Circus], the (hopefully futureproof) CI I am
working on at [Manic Systems], where the difference between "wait for
evaluation" and "ask the already evaluated world what changed" is critical for
our use case. The latter is a much nicer sentence to live inside---it's also a
nicer way of doing what Hydra does, and more reliable too.

## Workers That Aren't Here (They Are On Strike)

> A worker is an odd little thing that I decided mirror from Circus' distributed

Reviewer: ```suggestion
A worker is an odd little thing that I decided to mirror from Circus' distributed
```

---
title: "Nix Evaluation Is a Scheduling Problem"

> description: "Why I built Evix, and Rambling On Nix Evaluation"

Reviewer: are you sure you want your _description_ field to be in titlecase? but if you do, build -> Build

backend, Evix can live as a native sidecar or daemon behind whatever HTTP or
WebSocket API you actually want. If you want browser WASM, bad news: the current
backend is native by design. It links libnix, opens a Nix store, uses Unix
subprocesses, and speaks Unix sockets and TCP. An in-browser evaluator would
need a different backend, probably one we control entirely. Sad, I know. Alas,
that's what we have. However it's not the end.

### What Evix Solves

The problem Evix solves is not "Nix evaluation exists." The problem is that
serious tools around Nix need evaluation to become a reusable service boundary,
and not a ceremony performed by a subprocess every time. For CI specifically,
the useful object is a graph that can be diffed between commits. For your fleet
of NixOS configurations, it is a graph that can be filtered by system, prefix,
name or `.drv` path. For a watch command, it is a previous graph and a new graph
with errors attached to the comparison instead of hidden in a terminal
scrollback. For distributed evaluation, it is a queue of independent attribute
paths that local and remote workers can pull from without changing what the

> final graph means. For an embedding application, it is a typed API that does not

Reviewer: not sure, but did you mean to say embedded as opposed to embedding here?

ask the caller to summon a shell and interpret stdout as ontology. Thus, this
I've built Evix. Circus needs to evaluate jobsets often, keep them around, ask
what changed, and eventually hand builds to the right place. I do not want that
CI to spend its life recomputing the same graph because the previous process
threw away the one thing I cared about. I do not want an `aarch64` board to be
represented by an `x86_64` laptop. I do not want one broken package to make the
rest of the job graph unknowable. I do not want the evaluator and the builder to
become one swollen process just because both words appear near each other in a

> README. Yada yada. My point is that Evix draws the line where I needed it.

Reviewer: ```suggestion
README. Yada yada. My point is that Evix draws the line where I need it to.
```
might be better >,>

Evaluation discovers derivations. Building is somebody else's problem, usually
Nix itself or [Circus]. Building is stateful, resource-hungry, cache-sensitive,
failure-prone in completely different ways, and full of policy. Evaluation is
already enough of a problem, and I have 99 problems. Evaluation is 99 of them.

### Where It Stops

Evix is rather young, and there are real ceilings. If you end up trying Evix and
get rather disappointed, I ask that you attribute to its lack of maturity. I've
recently released 2.0.0 to revise its model in its entirety and it was a much
larger refactor than I feel I was equipped to deal with. Still, it turned out
quite well and Circus' own CI test suite is now green across the board. That

> said I'll be amiss if I don't mention the issues such as that a worker can still

Reviewer: ```suggestion
said, I'll be amiss if I don't mention the issues such as that a worker can still
```

die inside one awful force. Remote workers need compatible stores and trusted
networking; expose the TCP service over a tunnel or VPN, not to the entire
internet like you are trying to speedrun regret. System routing is also rather
constrained by the fact that the system is discovered after evaluation. The
daemon keeps warm sessions, but it is not a distributed database and should not
be asked to cosplay as one. THOUGH, it may be possible in the future. Last but

> not least remember that the C API is stable enough to build on, but Nix is still

Reviewer: ```suggestion
not least, remember that the C API is stable enough to build on, but Nix is still
```
